### DISCUSSION TITLE
How to using multi-stage build for Docker image with pnpm

### DISCUSSION THREAD

**Discussion by Planxnx** [pnpm/pnpm#5376] [Q&A] [answered]
I'v tried to mount Content-addressable store for sharing the cache between the stages, 
but I've got too many errors like this.

```
#16 1.811 Falling back to copying packages from store
#16 1.817  WARN  EXDEV: cross-device link not permitted, link '/root/.local/share/pnpm/store/v3/files/5e/082f14af74086bacffa6babb58741ece4080e26a84c915169e85ec95ed607cff3c302088ad7225a778ee34e24e5d837e22b02161cd30b3b6272bec1fd22d56' -> '/app/node_modules/.pnpm/express@4.18.1/node_modules/_tmp_1_0196054c05c40ebc538d930aab087189/LICENSE'
#16 1.817 Falling back to copying packages from store
#16 1.822  WARN  EXDEV: cross-device link not permitted, link '/root/.local/share/pnpm/store/v3/files/eb/c137c01c13da9d90fd411729b10cb15e6b7ea3f31a6b9dd23d0a769aefe61cbea7368c36698b0d562e85dca08ebcaa59996f9653295a1c79252cba3da3d3ad' -> '/app/node_modules/.pnpm/body-parser@1.20.0/node_modules/_tmp_1_9a51c710ed86958d85a3026561f85051/LICENSE'
```

What is the correct way to share the cache between the stages with out copy content-addressable store ?

this my Dockerfile
```
FROM node:16 as builder
WORKDIR /app

# Install package-manager
RUN corepack enable pnpm

# Install dependencies and mount Content-addressable store for multi-stage build
COPY .npmrc package.json pnpm-lock.yaml ./
RUN --mount=type=cache,id=pnpm-store,target=/root/.local/share/pnpm/store/v3 \
	pnpm install

# Build
COPY ./ ./
RUN pnpm build

FROM node:16-alpine
WORKDIR /app

# Install package-manager
RUN corepack enable pnpm

# Install only production dependencies and mount Content-addressable store to use downloaded dependencies
COPY --from=builder /app/.npmrc /app/package.json /app/pnpm-lock.yaml ./
RUN --mount=type=cache,id=pnpm-store,target=/root/.local/share/pnpm/store/v3 \
	pnpm install --prod

COPY --from=builder /app/dist/index.js ./index.js

CMD node index.js
```

**Comment by Planxnx** [comment, marked as answer]
I found another way to reduce image size, but is there other way?

```
FROM node:16 as builder
WORKDIR /app

RUN corepack enable pnpm

COPY pnpm-lock.yaml ./
RUN pnpm fetch

COPY .npmrc package.json ./
RUN pnpm install --offline

COPY ./ ./
RUN pnpm build

RUN pnpm prune --prod

FROM node:16-alpine
WORKDIR /app

COPY --from=builder /app/.npmrc /app/package.json /app/pnpm-lock.yaml ./
COPY --from=builder /app/node_modules ./node_modules
COPY --from=builder /app/dist ./

CMD node index.js
```

**Reply by kasir-barati** [reply]
So basically we cannot have something like this?
```Dockerfile
FROM node:20.10.0-slim as base
ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
# Install package-manager
RUN corepack enable pnpm
WORKDIR ${APP_WORKDIR}
RUN curl -f https://get.pnpm.io/v6.16.js | node - add --global pnpm
COPY pnpm-lock.yaml ${APP_WORKDIR}
RUN pnpm fetch

FROM base AS prod-deps
RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install -r --offline --prod


FROM base as build
RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install -r --offline
RUN pnpm run build


FROM base
COPY --from=prod-deps ${APP_WORKDIR}/node_modules ${APP_WORKDIR}/node_modules
COPY --from=build ${APP_WORKDIR}/dist ${APP_WORKDIR}/dist
COPY . ${APP_WORKDIR}
EXPOSE ${APP_PORT}
CMD [ "pnpm", "start" ]
```

**Reply by Planxnx** [reply]
> So basically we cannot have something like this?
> 
> ```Dockerfile
> 
> FROM node:20.10.0-slim as base
> 
> ENV PNPM_HOME="/pnpm"
> 
> ENV PATH="$PNPM_HOME:$PATH"
> 
> # Install package-manager
> 
> RUN corepack enable pnpm
> 
> WORKDIR ${APP_WORKDIR}
> 
> RUN curl -f https://get.pnpm.io/v6.16.js | node - add --global pnpm
> 
> COPY pnpm-lock.yaml ${APP_WORKDIR}
> 
> RUN pnpm fetch
> 
> 
> 
> FROM base AS prod-deps
> 
> RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install -r --offline --prod
> 
> 
> 
> 
> 
> FROM base as build
> 
> RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install -r --offline
> 
> RUN pnpm run build
> 
> 
> 
> 
> 
> FROM base
> 
> COPY --from=prod-deps ${APP_WORKDIR}/node_modules ${APP_WORKDIR}/node_modules
> 
> COPY --from=build ${APP_WORKDIR}/dist ${APP_WORKDIR}/dist
> 
> COPY . ${APP_WORKDIR}
> 
> EXPOSE ${APP_PORT}
> 
> CMD [ "pnpm", "start" ]
> 
> ```

If we can't use slim/alpine node image in build stage(E.g. use node:20.0 for base), the output image will still be very large.

I want to use multi-stage because I want the output image to be small size, that's why I've use node:16 for build stage and node:16-alpine for run application.

But your solution that's interesting, If it works if output stage is supported to use alpine image 🙏 


Note: I can't mount store and run install deps, so I've to workaround like this solution.

**Reply by kasir-barati** [reply]
@Planxnx my issue was how I was saying to install packages offline:

**Wrong** `RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install -r --offline`
**Correct** `RUN pnpm install -r --offline`

Now you can add `--prod` flag if you need it. Also IDK why, still wip, but for some reason (I think more related to how docker works) I needed to execute `pnpm install -r --offline` in the build stage.

# Final Dockerfile
This is my dev docker file. As such I really thought copying entire source code into the image should not raise any issue, not for the foreseeable future that I can see 😁.

But of. course for the prod you might want to change it. Then there you need to copy package.json, and tsconfig.json (Not sure what else but in my case it was not working without tsconfig file 😕) too
```dockerfile
ARG APP_PORT
ARG APP_WORKDIR


FROM node:20.10.0-slim AS base
ARG APP_WORKDIR
ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
RUN corepack enable
WORKDIR ${APP_WORKDIR}
COPY pnpm-lock.yaml ${APP_WORKDIR}
RUN pnpm fetch


FROM base AS build
ARG APP_WORKDIR
COPY . ${APP_WORKDIR}
RUN pnpm install -r --offline
RUN pnpm run build


FROM base
ARG APP_PORT
ARG APP_WORKDIR
COPY --from=build ${APP_WORKDIR} ${APP_WORKDIR}
EXPOSE ${APP_PORT}
CMD [ "pnpm", "start" ]
```
